File: CPP_SOLUTIONS/131.cpp
class Solution {
public:
    vector<vector<string>> res;
    vector<string> tmp;
    bool Palindrome(string s) {
        for(int i = 0; i < s.size() / 2; i++) if(s[i] != s[s.size() - 1 - i]) return 0;
        return 1;
    } 
    void dfs(string s) {
        if(s.size() == 0) {
            res.push_back(tmp);
            return;
        }
        for(int i = 1; i <= s.size(); i++) {
            if(Palindrome(s.substr(0, i))) {
                //cout << s.substr(0, i) << endl;
                tmp.push_back(s.substr(0, i));
                dfs(s.substr(0 + i));
                tmp.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        res.clear();
        tmp.clear();
        if(s == "") return res;
        dfs(s);
        return res;
    }
};

File: PYTHON_SOLUTIONS/131.py
class Solution(object):
    res = []
    stack = []
    
    def isPalindrome(self, s):
        #print(s)
        for i in range(len(s)/2):
            if s[i] != s[-i-1]:
                return False
        return True
    
    def dfs(self, s, idx):
        n = len(s)
        #print idx
        if idx == n:
            #print self.stack
            self.res.append(list(self.stack))
            return
        for l in range(1, len(s) - idx + 1):
            #print(s[idx: idx + l], self.isPalindrome(s[idx: idx + l]))
            if self.isPalindrome(s[idx: idx + l]):
                #print(s[idx: idx + l])
                self.stack.append(s[idx: idx + l])
                self.dfs(s, idx + l)
                self.stack.pop()
    
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        self.res = []
        self.stack = []
        self.dfs(s, 0)
        return self.res
        
